### PR TITLE
[#2] New Policy Example - Logical Quotas for Object Count and Data Size

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: irods_policy_logical_quotas.re [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+# a background process to keep things in sync could remedy this issue.
+#
+# ------
+#
+# Getting Started:
+#
+# To enable logical quotas, you first need to be running the Native Rule Engine Plugin.
+# If you have that, all you need to do is run the following command:
+#
+#   irule -r <native_rule_language_instance> 'logical_quotas_init(<collection>, <max_object_count>, <max_data_size_in_bytes>)' null null
```
Comment (trel) [inline irods_policy_logical_quotas.re]: maximum_object_count and maximum_data_size_in_bytes

File: irods_policy_logical_quotas.re [hunk 2]
```diff
@@ -0,0 +1,570 @@
...
+# The default error code returned when a logical quota is exceeded.
+SYS_INVALID_INPUT_PARAM { -130000; }
+
+# Returns *name prefixed with "irods::logical_quotas::".
+#
+# Parameters:
+# - *name: The string to prefix.
+logical_quotas_key(*name)
+{
+    "irods::logical_quotas::*name";
```
Comment (trel) [inline irods_policy_logical_quotas.re]: does this work if they change lines 41-44?   the hard-coded prefix breaks the abstraction?

File: irods_policy_logical_quotas.re [hunk 3]
```diff
@@ -0,0 +1,570 @@
...
+logical_quotas_get_parent_collection(*path)
+{
+    *parent = "";
+
+    if (*path != "" && *path != "/") {
+        *ec = errormsg(msiSplitPath(*path, *parent, *_), *msg);
+        logical_quotas_fail_if_split_path_error(*ec, *path);
+    }
+
+    *parent
```
Comment (trel) [inline irods_policy_logical_quotas.re]: semicolon?

File: irods_policy_logical_quotas.re [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+    *ccount = logical_quotas_current_object_count_key();
+    *csize = logical_quotas_current_data_size_in_bytes_key();
+
+    *kvp.*ccount = str(int(*kvp.*ccount) + *num_objects);
+    *kvp.*csize = str(int(*kvp.*csize) + *size_in_bytes);
+}
+
+# The general pattern for the PEPs is as follows:
+# - Pre PEPs are used to detect if a certain operation will violate the rules and if so, stop immediately.
+#   Pre PEPs do NOT modify the totals in anyway. They may attach additional metadata needed by the Post PEP.
```
Comment (trel) [inline irods_policy_logical_quotas.re]: - Pre PEPs, missing hyphen?

File: irods_policy_logical_quotas.re [hunk 5]
```diff
@@ -0,0 +1,570 @@
...
+            if (*obj_type == "-d") {
+                logical_quotas_fail_if_max_count_violation(*kvp, 1);
+                logical_quotas_fail_if_max_size_violation(*kvp, logical_quotas_get_data_object_size(*INPUT.src_obj_path));
+            }
+            else if (*obj_type == "-c") {
+                logical_quotas_compute_object_count_and_size(*INPUT.src_obj_path, *count, *size);
+                logical_quotas_fail_if_max_count_violation(*kvp, *count);
+                logical_quotas_fail_if_max_size_violation(*kvp, *size);
+            }
+        }
```
Comment (trel) [inline irods_policy_logical_quotas.re:559]: what's the else() case?   error?

File: irods_policy_logical_quotas.re [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+                *ec = errormsg(msiSetKeyValuePairsToObj(*kvp, *dst_parent_path, "-C"), *msg);
+                logical_quotas_fail_if_set_key_value_pairs_error(*ec, *dst_parent_path);
+            }
+            else if (*obj_type == "-c") {
+                logical_quotas_compute_object_count_and_size(*INPUT.src_obj_path, *count, *size);
+                logical_quotas_add_objects_and_size(*kvp, *count, *size);
+                *ec = errormsg(msiSetKeyValuePairsToObj(*kvp, *dst_parent_path, "-C"), *msg);
+                logical_quotas_fail_if_set_key_value_pairs_error(*ec, *dst_parent_path);
+            }
+        }
```
Comment (trel) [inline irods_policy_logical_quotas.re:590]: else case...?  just continue?

**Review by trel — COMMENTED** [review]
this is really good.    a nice generic framework for quotas.

could replace the existing implementation....(!)

File: irods_policy_logical_quotas.re (hunk 5 above)
Comment (korydraughn) [reply irods_policy_logical_quotas.re:559]: is it possible for a type other than **collection** or **data object** to be returned?  it seems that the only way for something to go wrong is if the REPF passed something other than a path.

Comment (trel) [reply irods_policy_logical_quotas.re:559]: https://github.com/irods/irods/blob/4-2-stable/server/core/src/objMetaOpr.cpp#L285

Comment (korydraughn) [reply irods_policy_logical_quotas.re:559]: I'm thinking maybe it just continues, although if some other type appears in one of these peps, that would mean there is a bigger problem, right?

Comment (trel) [reply irods_policy_logical_quotas.re:559]: yeah, let's have have an else() that says 'should never get here' and reports things are not in a good place...

File: irods_policy_logical_quotas.re (hunk 1 above)
Comment (korydraughn) [reply irods_policy_logical_quotas.re]: updated.

File: irods_policy_logical_quotas.re (hunk 2 above)
Comment (korydraughn) [reply irods_policy_logical_quotas.re]: adjusted to allow changing the namespace.

File: irods_policy_logical_quotas.re (hunk 3 above)
Comment (korydraughn) [reply irods_policy_logical_quotas.re]: semicolon added.

File: irods_policy_logical_quotas.re (hunk 4 above)
Comment (korydraughn) [reply irods_policy_logical_quotas.re]: added hyphens.

File: irods_policy_logical_quotas.re (hunk 5 above)
Comment (korydraughn) [reply irods_policy_logical_quotas.re:559]: added else case.

File: irods_policy_logical_quotas.re (hunk 6 above)
Comment (korydraughn) [reply irods_policy_logical_quotas.re:590]: added else case.

**Pull request merged**